vmdb/app/models/metric/ci_mixin/capture/openstack_base.rb

```diff
@@ -37,134 +37,364 @@ def perf_release_openstack
+            # skipped.
+            multi_counter_aligned_start = last_period
           else
-            # the aggregate is still not divisible by 20, grab the next stat
-            next
+            # Not beginning, the data are corrupted or missing. In order to avoid holes in saved data, this period
+            # will be filled by data of the next periods.
+            $log.warn("#{log_header} id:[#{name}] Distance of the multiple streams of data is invalid. It "\
+                      "exceeded half of the Ceilometer collection period in "\
+                      "#{multi_counter_aligned_start_guard} - #{last_period} for counters" \
+                      "#{i[:openstack_counters]} It can be caused by different pipeline configuration "\
+                      "period for each related sample. Or you need to scale Ceilometer because the samples"\
+                      "collection is overloaded.")
           end
+          multi_counter_metrics      = {}
+          last_multi_counter_metrics = {}
+          # Moving to guard to another period cause the old one had incomplete data for all counters. We will
+          # try to collect all data for all counters in next guard period.
+          multi_counter_aligned_start_guard = last_period
         end
 
-        # For (temporary) symmetry with VIM API we create 20-second intervals.
-        (last_ts + 20.seconds..ts).step_value(20.seconds).each do |ts_item|
-          counter_values_by_ts.store_path(ts_item.iso8601, i[:vim_style_counter_key], value)
+        i[:openstack_counters].each { |c| last_metrics[c] = metrics_by_counter_name.fetch_path(c, last_period) }
+        all_last_metrics_available = process_multi_counter_metrics(i, last_metrics, last_multi_counter_metrics)
+
+        if !all_last_metrics_available
+          # Resetting multi_counter_aligned_end, which will be initialized to period in next period
+          multi_counter_aligned_end = nil
+          # All last_multi_counter data are not available, lets move to another period and try to capture them
+          next
+        else
+          # Move guard when last_period is complete
+          multi_counter_aligned_start_guard = period
+        end
+      end
+
+      if period - multi_counter_aligned_start_guard > data_collecting_period / 2
+        # The data are corrupted or missing. In order to avoid holes in saved data, this period will be filled
+        # by data of next periods.
+        $log.warn("#{log_header} id:[#{name}] Distance of the multiple streams of data is invalid. It "\
+                  "exceeded half of the Ceilometer collection period in "\
+                  "(#{multi_counter_aligned_start_guard}, #{last_period}> for counters"\
+                  "#{i[:openstack_counters]} It can be caused by different pipeline configuration "\
+                  "period for each related sample. Or you need to scale Ceilometer because the samples"\
+                  "collection is overloaded.")
```

Comment:
Having this identical message twice makes me think this needs to be refactored or moved into a method or something.